lib/corefoundation/base.rb

```diff
@@ -1,101 +1,37 @@
-  attach_function :release, 'CFRelease', [:cftyperef], :void
-  attach_function :retain, 'CFRetain', [:cftyperef], :cftyperef
-  attach_function 'CFEqual', [:cftyperef, :cftyperef], :char
-  attach_function 'CFHash', [:cftyperef], :cfhashcode
-  attach_function 'CFCopyDescription', [:cftyperef], :cftyperef
-  attach_function 'CFGetTypeID', [:cftyperef], :cftypeid
-
-  # The base class for all of the wrapper classes 
-  #
-  # @abstract
-  class Base  
-    @@type_map = {}
-
-    # @private
-    class Releaser
-      def initialize(ptr)
-        @address  = ptr.address
-      end
-
-      def call *ignored
-        if @address != 0
-          CF.release(@address)
-          @address = 0
-        end
-      end
+    # Raises an exception if the argument does not inherit from CF::Base
+    #
+    # @param cftyperef object to check
+    def self.check_cftype(cftyperef)
+      raise TypeError, "#{cftyperef.inspect} is not a cftype" unless cftyperef.is_a?(CF::Base)
     end
 
-    class << self
-      # Raises an exception if the argument does not inherit from CF::Base
-      #
-      # @param cftyperef object to check
-      def check_cftype(cftyperef)
-        raise TypeError, "#{cftyperef.inspect} is not a cftype" unless cftyperef.is_a?(CF::Base)
-      end
-
-      # Wraps an FFI::Pointer with the appropriate subclass of CF::Base
-      # If the pointer is not a CFTypeRef behaviour is undefined
-      # @param [FFI::Pointer] cftyperef Object to wrap
-      # @return A wrapper object inheriting from CF::Base
-      def typecast(cftyperef)
-        klass = klass_from_cf_type cftyperef
-        klass.new(cftyperef)
-      end
-
-      private
-      def klass_from_cf_type cftyperef
-        klass = @@type_map[CF.CFGetTypeID(cftyperef)]
-        if !klass
-          raise TypeError, "No class registered for cf type #{cftyperef.inspect}"
-        end
-        klass
-      end
-
-      def register_type(type_name)
-        CF.attach_function "#{type_name}GetTypeID", [], :cftypeid
-        @@type_map[CF.send("#{type_name}GetTypeID")] = self
-      end
+    # Wraps an FFI::Pointer with the appropriate subclass of CF::Base
+    # If the pointer is not a CFTypeRef behaviour is undefined
+    #
+    # @param [FFI::Pointer] cftyperef Object to wrap
+    # @return A wrapper object inheriting from CF::Base
+    def self.typecast(cftyperef)
+      klass = klass_from_cf_type cftyperef
+      klass.new(cftyperef)
+    end
 
+    # @param [FFI::Pointer] pointer The pointer to wrap
+    def initialize(pointer)
+      @ptr = FFI::Pointer.new(pointer)
+      ObjectSpace.define_finalizer(self, self.class.finalize(ptr))
     end
 
-    # 
-    #
-    # @param [FFI::Pointer] ptr The pointer to wrap
-    def initialize(ptr)
-      @ptr = FFI::Pointer.new(ptr)
+    # @param [FFI::Pointer] pointer to the address of the object
+    def self.finalize(pointer)
+      proc { CF.release(pointer.address) unless pointer.address.zero }
```

Comment:
To obliviate all local binding concerns here, we can have a class object instead, as finalizer

Reply:
I think this might  be broken - this should be `pointer.address.zero?`, not `pointer.address.zero`, and this should be `CF.release(pointer)`, not `CF.release(pointer.address)`. You can see this easily enough by running this code outside of a finalizer - finalizers swallow exception unfortunately.

This could do with some really solid unit tests, using CFGetRetainCount to check that things get released appropriately - mea culpa for never having written those in the past

Reply:
@fcheung That's interesting. I didn't know this. I can probably write some tests around this. I'll create a separate PR for this change though since this PR is getting bigger than I anticipated and probably best to have the discussion in a dedicated place too.

Reply:
fixed it in this PR, would be tracking tests in a separate PR though.

Reply:
Can we verify that there is no binding of self/metaclass in proc in any case?

Reply:
We can continue the discussions related to memory management in #12

Reply:
Could you then revert the code as well, why merge something that's being addressed in another issue?